presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java

```diff
@@ -578,4 +578,22 @@ default ConnectorTableProperties getTableProperties(ConnectorSession session, Co
 
         return new ConnectorTableProperties(layouts.get(0).getTableLayout());
     }
+
+    /**
+     * Attempt to push down the provided limit into the table.
+     * <p>
+     * Connectors can indicate whether they don't support limit pushdown or that the action had no effect
+     * by returning {@link Optional#empty()}.
+     * <p>
+     * If the connector could benefit from the information but can't guarantee that it will be able to produce
+     * fewer rows than the provided limit, it should return a non-empty result containing a new handle for the
+     * derived table and the "limit guaranteed" flag set to false.
```

Comment:
Do we need to mention that it is important for the connector to only return a new handle instance if something changed?

Reply:
It's there, but probably not clear enough. I'll make more emphasis on that.